src/esrs/aggregate.rs

```diff
@@ -20,69 +20,69 @@ pub trait Identifier {
+/// The Aggregate trait is responsible for validating commands, mapping commands to events, and applying
+/// events onto the state.
 ///
-/// Unless you need to perform side effects as part of your command handling/verification you should implement the
-/// safer `Aggregate` trait instead.
-#[async_trait]
-pub trait AggregateManager: Identifier {
+/// An Aggregate should be able to derive its own state from nothing but its initial configuration, and its
+/// event stream. Applying the same events, in the same order, to the same aggregate, should always yield an
+/// identical aggregate state.
+///
+/// This trait is purposefully _synchronous_. If you are implementing this trait, your aggregate
+/// should not have any side effects. If you need additional information to handle commands correctly, then
+/// consider looking up that information and placing it in the command.
+pub trait Aggregate {
     type State: Default + Clone + Debug + Send + Sync;
     type Command: Send + Sync;
     type Event: Serialize + DeserializeOwned + Send + Sync;
     type Error: Send + Sync;
 
-    /// Returns the event store, configured for the aggregate
-    fn event_store(&self) -> &(dyn EventStore<Self::Event, Self::Error> + Send + Sync);
-
-    /// This function applies the event onto the aggregate and returns a new one, updated with the event data
-    fn apply_event(id: &Uuid, state: Self::State, event: &StoreEvent<Self::Event>) -> Self::State;
+    /// Handles, validate a command and emits events.
+    fn handle_command(state: &AggregateState<Self::State>, cmd: Self::Command)
+        -> Result<Vec<Self::Event>, Self::Error>;
 
-    /// Validation should reject any command is inconsistent with the current aggregate state, or would result
-    /// in one or more events that could not be applied onto the aggregate state.
-    fn validate_command(aggregate_state: &AggregateState<Self::State>, cmd: &Self::Command) -> Result<(), Self::Error>;
+    /// Updates the aggregate state using the new event. This assumes that the event can be correctly applied
+    /// to the state.
+    ///
+    /// If this is not the case, this function is allowed to panic.
+    fn apply_event(state: Self::State, payload: &Self::Event) -> Self::State;
+}
 
-    async fn do_handle_command(
-        &self,
-        aggregate_state: AggregateState<Self::State>,
-        cmd: Self::Command,
-    ) -> Result<AggregateState<Self::State>, Self::Error>;
+/// The AggregateManager is responsible for coupling the Aggregate with a Store, so that the events
+/// can be persisted when handled, and the state can be reconstructed by loading and apply events sequentially.
+///
+/// It comes batteries-included, as you only need to implement the `event_store` getter. The basic API is:
+/// 1. execute_command
+/// 2. load
+/// The other functions are used internally, but can be overridden if needed.
+#[async_trait]
+pub trait AggregateManager: Aggregate + Identifier {
```

Comment:
Probably we don't need the `Identifier` bound here, as the `PgStore<T: Identifier>` requires it.
```suggestion
pub trait AggregateManager: Aggregate {
```

Reply:
Ok but then who implements Identifier?
Anyway i was thinking about remove the `Identifier` trait and just add a function `name` in the `Aggregate`/`AggregateManager`. Wdyt?

Reply:
I think the function would be even better.